internal/adaptorsigs/adaptor.go

```diff
@@ -0,0 +1,532 @@
+	// 4. s = int(sig[32:64]); fail if s >= n.
+	// 5. e = int(tagged_hash("BIP0340/challenge", bytes(r) || bytes(P) || M)) mod n.
+	// 6. R = s*G - e*P - T
+	// 7. Fail if is_infinite(R)
+	// 8. Fail if not hash_even_y(R)
+	// 9. Fail is x(R) != r.
+	// 10. Return success iff not failure occured before reachign this
+	// point.
+
+	// Step 1.
+	//
+	// Fail if m is not 32 bytes
+	if len(hash) != scalarSize {
+		str := fmt.Sprintf("wrong size for message (got %v, want %v)",
+			len(hash), scalarSize)
+		return errors.New(str)
+	}
+
+	// Step 2.
+	//
+	// P = lift_x(int(pk))
+	//
+	// Fail if P is not a point on the curve
+	pubKey, err := schnorr.ParsePubKey(pubKeyB)
+	if err != nil {
+		return err
+	}
+	if !pubKey.IsOnCurve() {
+		str := "pubkey point is not on curve"
+		return errors.New(str)
+	}
+
+	// Step 3.
+	//
+	// Fail if r >= p
+	//
+	// Note this is already handled by the fact r is a field element.
+
+	// Step 4.
+	//
+	// Fail if s >= n
+	//
+	// Note this is already handled by the fact s is a mod n scalar.
+
+	// Step 5.
+	//
+	// e = int(tagged_hash("BIP0340/challenge", bytes(r) || bytes(P) || M)) mod n.
+	var rBytes [32]byte
+	sig.r.PutBytesUnchecked(rBytes[:])
+	pBytes := schnorr.SerializePubKey(pubKey)
+	commitment := chainhash.TaggedHash(
+		chainhash.TagBIP0340Challenge, rBytes[:], pBytes, hash,
+	)
+
+	var e btcec.ModNScalar
+	e.SetBytes((*[32]byte)(commitment))
+
+	// Negate e here so we can use AddNonConst below to subtract the s*G
+	// point from e*P.
+	e.Negate()
+
+	// Step 6.
+	//
+	// R = s*G - e*P - T
+	var P, R, sG, eP, encryptedR btcec.JacobianPoint
+	pubKey.AsJacobian(&P)
+	btcec.ScalarBaseMultNonConst(&sig.s, &sG)
+	btcec.ScalarMultNonConst(&e, &P, &eP)
+	btcec.AddNonConst(&sG, &eP, &R)
+	tInv := sig.t
+	tInv.Y.Negate(1)
+	secp256k1.AddNonConst(&R, &tInv, &encryptedR)
+
+	// Step 7.
+	//
+	// Fail if R is the point at infinity
+	if (encryptedR.X.IsZero() && encryptedR.Y.IsZero()) || encryptedR.Z.IsZero() {
+		str := "calculated R point is the point at infinity"
+		return errors.New(str)
+	}
+
+	// Step 8.
+	//
+	// Fail if R.y is odd
+	//
+	// Note that R must be in affine coordinates for this check.
+	encryptedR.ToAffine()
+	if encryptedR.Y.IsOdd() {
+		str := "calculated R y-value is odd"
+		return errors.New(str)
+	}
+
+	// Step 9.
+	//
+	// Verified if R.x == r
+	//
+	// Note that R must be in affine coordinates for this check.
+	if !sig.r.Equals(&encryptedR.X) {
+		str := "calculated R point was not given R"
+		return errors.New(str)
+	}
+
+	// Step 10.
+	//
+	// Return success iff not failure occurred before reaching this
+	return nil
+}
+
+// Verify checks that the adaptor signature will result in a valid signature
+// when decrypted with the tweak.
+func (sig *AdaptorSignature) Verify(hash []byte, pubKey *secp256k1.PublicKey) error {
+	if sig.pubKeyTweak {
+		return fmt.Errorf("only private key tweaked adaptors can be verified")
+	}
+	pubKeyBytes := schnorr.SerializePubKey(pubKey)
```

Comment:
This ensures that the y coordinate is always even.

Reply:
Are you sure?  Unless `btcec` is now doing something make that case, that isn't true.  I'll take a look here in a bit to double check that.

Reply:
So, tracing the code:

1) [`btcec/shnorr.SerializePubKey`](https://github.com/btcsuite/btcd/blob/2b53ed198955ac40fe5b3ce4a854e9f5bfa68258/btcec/schnorr/pubkey.go#L47C16-L47C35) calls `btcec/v2.PublicKey.SerializeCompressed`
2) [`btcec/v2.PublicKey`](https://github.com/btcsuite/btcd/blob/2b53ed198955ac40fe5b3ce4a854e9f5bfa68258/btcec/pubkey.go#L42) is just a type alias to `dcrec/secp256k1/v4.PublicKey`
3) `btcec/v2` is [using version 4.0.1 of `dcrec/secp256k1/v4`](https://github.com/btcsuite/btcd/blob/2b53ed198955ac40fe5b3ce4a854e9f5bfa68258/btcec/go.mod#L8)
4) Finally, [dcrec/secp256k1/v4@v4.0.1.PublicKey.SerializeCompressed](https://github.com/decred/dcrd/blob/dcrec/secp256k1/v4.0.1/dcrec/secp256k1/pubkey.go#L198-L210) does **NOT** (nor should it) do anything to enforce the evenness of the y coordinate.

So, this does not look correct to me.

Reply:
https://github.com/btcsuite/btcd/blob/2b53ed198955ac40fe5b3ce4a854e9f5bfa68258/btcec/schnorr/pubkey.go#L48C2-L48C8

Over here, the first byte which specifies either even or odd is removed. Then, when it's parsed, it is assumed to be even.

Reply:
Right, I saw that it removes the byte, but that doesn't actually ensure it was originally even at all.  I realize it will make it even during the decompression, but then it's no longer the same curve point which means that if it wasn't even to begin with and you do this, you'll just cause the verification to fail anyway.

Here is a simplified playground example showing that: https://go.dev/play/p/8eCL1-YW738.

Reply:
Turns out I made a big mistake, I copy pasted the btc adaptor sigs code from my PR in the adaptorsigs repo to here, and I assumed the replacement just worked because all of the tests passed. The redeem transactions were actually not being mined, but the balance was still being updated.

Reply:
That looks better.